pkg/rtc/participant.go

```diff
@@ -1355,112 +1067,55 @@ func (p *ParticipantImpl) handleDataMessage(kind livekit.DataPacket_Kind, data [
+func (p *ParticipantImpl) onPrimaryTransportInitialConnected() {
+	if !p.hasPendingMigratedTrack() && p.MigrateState() == types.MigrateStateSync {
+		p.SetMigrateState(types.MigrateStateComplete)
 	}
 }
 
-func (p *ParticipantImpl) handlePrimaryStateChange(state webrtc.PeerConnectionState) {
-	if state == webrtc.PeerConnectionStateConnected {
-		if !p.firstConnected.Swap(true) {
-			p.setDowntracksConnected()
-		}
-		prometheus.ServiceOperationCounter.WithLabelValues("ice_connection", "success", "").Add(1)
-		if !p.hasPendingMigratedTrack() && p.MigrateState() == types.MigrateStateSync {
-			p.SetMigrateState(types.MigrateStateComplete)
-		}
-		p.incActiveCounter()
-	} else if state == webrtc.PeerConnectionStateFailed {
-		p.handleConnectionFailed(true)
-
-		// clients support resuming of connections when websocket becomes disconnected
-		p.closeSignalConnection()
-
-		// detect when participant has actually left.
-		go func() {
-			p.lock.Lock()
-			if p.disconnectTimer != nil {
-				p.disconnectTimer.Stop()
-				p.disconnectTimer = nil
-			}
-			p.disconnectTimer = time.AfterFunc(disconnectCleanupDuration, func() {
-				p.lock.Lock()
-				p.disconnectTimer.Stop()
-				p.disconnectTimer = nil
-				p.lock.Unlock()
-
-				if p.isClosed.Load() || p.State() == livekit.ParticipantInfo_DISCONNECTED {
-					return
-				}
-				primaryPC := p.publisher.pc
-				if p.SubscriberAsPrimary() {
-					primaryPC = p.subscriber.pc
-				}
-				if primaryPC.ConnectionState() != webrtc.PeerConnectionStateConnected {
-					p.params.Logger.Infow("closing disconnected participant")
-					_ = p.Close(true, types.ParticipantCloseReasonPeerConnectionDisconnected)
-				}
-			})
-			p.lock.Unlock()
-		}()
-
-	}
+func (p *ParticipantImpl) onPrimaryTransportFullyEstablished() {
+	p.updateState(livekit.ParticipantInfo_ACTIVE)
 }
 
-// for the secondary peer connection, we still need to handle when they become disconnected
-// instead of allowing them to silently fail.
-func (p *ParticipantImpl) handleSecondaryStateChange(state webrtc.PeerConnectionState) {
-	if state == webrtc.PeerConnectionStateFailed {
-		p.handleConnectionFailed(false)
+func (p *ParticipantImpl) onAnyTransportFailed() {
+	// clients support resuming of connections when websocket becomes disconnected
+	p.closeSignalConnection()
 
-		// clients support resuming of connections when websocket becomes disconnected
-		p.closeSignalConnection()
+	// detect when participant has actually left.
+	p.lock.Lock()
+	if p.disconnectTimer != nil {
+		p.disconnectTimer.Stop()
+		p.disconnectTimer = nil
 	}
+	p.disconnectTimer = time.AfterFunc(disconnectCleanupDuration, func() {
+		p.lock.Lock()
+		p.disconnectTimer.Stop()
+		p.disconnectTimer = nil
+		p.lock.Unlock()
+
+		if p.isClosed.Load() || p.State() == livekit.ParticipantInfo_DISCONNECTED {
+			return
+		}
+		p.params.Logger.Infow("closing disconnected participant")
+		_ = p.Close(true, types.ParticipantCloseReasonPeerConnectionDisconnected)
+	})
+	p.lock.Unlock()
```

Comment:
This timer was set up to run only on primary transport. I changed it to run on any transport as we are closing signal connection on any transport failure. So, it made to run this on any transport failure. Also, if any transport fails, that would be bad. So, running this on any transport failure.